### PR TITLE
feat(core): compute camera near/far automatically

### PR DESCRIPTION
**Pull request by peppsac**
Adds a new feature to compute automatically min/max distance of displayed elements.
The near/far values are then deduced.

The advantage of this solution is we can disable the logarithmicDepthBuffer by default which improve itowns compatibility with lower end / mobile devices.

### CODE REVIEW

File: examples/3dtiles.html [hunk 1]
```diff
@@ -21,17 +21,20 @@
             // iTowns namespace defined here
             var viewerDiv = document.getElementById('viewerDiv');
 
-            var view = new itowns.GlobeView(viewerDiv, positionOnGlobe);
+            var view = new itowns.GlobeView(viewerDiv, positionOnGlobe, { noControls: true });
             view.camera.camera3D.near = 5;
             setupLoadingScreen(viewerDiv, view);
 
+            var controls = new itowns.FirstPersonControls(view, { focusOnClick: true });
```
Comment (gchoqueux) [inline examples/3dtiles.html]: why you change controls?

File: src/Core/MainLoop.js [hunk 2]
```diff
@@ -121,8 +121,22 @@ MainLoop.prototype._update = function _update(view, updateSources, dt) {
         engine: this.gfxEngine,
         scheduler: this.scheduler,
         view,
+        distance: {
```
Comment (gchoqueux) [inline src/Core/MainLoop.js]: you could comment on these new parameters, `distance` and `fastUpdateHint`

File: src/Core/MainLoop.js [hunk 3]
```diff
@@ -121,8 +121,22 @@ MainLoop.prototype._update = function _update(view, updateSources, dt) {
...
+            min: Infinity,
+            max: 0,
+        },
+        fastUpdateHint: undefined,
     };
 
+    const previousNear = view.camera.camera3D.near;
+    const previousFar = view.camera.camera3D.far;
+    view.camera.camera3D.near = 0.1;
```
Comment (gchoqueux) [inline src/Core/MainLoop.js]: the near and the far are reset here for processing requiring the camera?

and they are restored at the end of the process?

File: src/Process/3dTilesProcessing.js [hunk 4]
```diff
@@ -257,24 +257,36 @@ export function pre3dTilesUpdate(context, layer) {
...
+export function computeNodeSSE(context, camera, node) {
     node.distance = 0;
     if (node.boundingVolume.region) {
         boundingVolumeBox.copy(node.boundingVolume.region.box3D);
         boundingVolumeBox.applyMatrix4(node.boundingVolume.region.matrixWorld);
         node.distance = boundingVolumeBox.distanceToPoint(camera.camera3D.position);
+        // update distance min/max
+        const s = node.boundingVolume.region.box3D.getSize();
+        const maxComponent = Math.max(s.x, Math.max(s.y, s.z));
+        context.distance.min = Math.min(context.distance.min, node.distance - maxComponent);
```
Comment (gchoqueux) [inline src/Process/3dTilesProcessing.js]: the calculations of min and max can be done, only once, after the calculation of the `boudingVolume`

The name of the function must also take into account this new calculation

File: src/Process/PointCloudProcessing.js [hunk 5]
```diff
@@ -86,6 +86,15 @@ function markForDeletion(elt) {
     }
 }
 
+function updateMinMaxDistance(context, bbox) {
```
Comment (gchoqueux) [inline src/Process/PointCloudProcessing.js:89]: this isn't a calculation equivalent to those performed in 3Dtiles?

File: src/Core/MainLoop.js [hunk 6]
```diff
@@ -153,6 +168,16 @@ MainLoop.prototype._update = function _update(view, updateSources, dt) {
             view.execFrameRequesters(MAIN_LOOP_EVENTS.AFTER_LAYER_UPDATE, dt, this._updateLoopRestarted, geometryLayer);
         }
     }
+
+    if (context.distance.min != Infinity) {
```
Comment (gchoqueux) [inline src/Core/MainLoop.js:189]: could be a camera function,` camera.adjustNearFar(min, max)`

File: src/Process/3dTilesProcessing.js [hunk 7]
```diff
@@ -257,24 +257,36 @@ export function pre3dTilesUpdate(context, layer) {
 
 const boundingVolumeBox = new THREE.Box3();
 const boundingVolumeSphere = new THREE.Sphere();
-export function computeNodeSSE(camera, node) {
+export function computeNodeSSE(context, camera, node) {
```
Comment (gchoqueux) [inline src/Process/3dTilesProcessing.js]: context used only for distance, so the params could be with the `distance`.

File: src/Core/MainLoop.js [hunk 8]
```diff
@@ -153,6 +168,16 @@ MainLoop.prototype._update = function _update(view, updateSources, dt) {
...
+        // Multiply min/max by (0.9, 1.1) to be on the safe side
+        view.camera.camera3D.near = Math.max(0.1, 0.9 * context.distance.min);
```
Comment (gchoqueux) [inline src/Core/MainLoop.js:191]: As the `min` distance is calculated relative to the camera's position, close objects may be truncated by the near.
This min distance must be projected on the z-axis of the camera

File: src/Process/PointCloudProcessing.js [hunk 9]
```diff
@@ -137,87 +146,95 @@ export default {
...
-                        }
-                        elt.obj.boxHelper.visible = true;
-                        elt.obj.boxHelper.material.color.r = 1 - elt.sse;
-                        elt.obj.boxHelper.material.color.g = elt.sse;
+            if (!elt.visible) {
+                markForDeletion(elt);
+                return;
+            }
+
+            const distance = updateMinMaxDistance(context, bbox);
```
Comment (gchoqueux) [inline src/Process/PointCloudProcessing.js:175]: I would do a function that calculates the distance, and then with the distance, I'll refresh the context

**Review by gchoqueux — CHANGES_REQUESTED** [review]
by aggregating the displayable bboxes, to obtain the bbox of the scene, (with box3.union), the calculation of the distance min and max would be made only once

File: src/Core/MainLoop.js (hunk 3 above)
Comment (peppsac) [reply src/Core/MainLoop.js]: Yes.
Because the update functions may use the view frustum to determine visibility.
The near/far affects the result, so we reset them to ~0 and ~Infinity before the processing.
After the visibility process is over, we can define them to the correct values (correct = the minimum z-range that contains all visible objects).

File: src/Process/3dTilesProcessing.js (hunk 4 above)
Comment (peppsac) [reply src/Process/3dTilesProcessing.js]: > the calculations of min and max can be done, only once, after the calculation of the boudingVolume

I guess you meant "after the calculation of `boundingBoxVolume` for the boundingVolume.region and boundingVolume.box cases" ?

File: src/Process/PointCloudProcessing.js (hunk 5 above)
Comment (peppsac) [reply src/Process/PointCloudProcessing.js:89]: Yep

File: src/Core/MainLoop.js (hunk 6 above)
Comment (peppsac) [reply src/Core/MainLoop.js:189]: Yeah, but in this case it would be accessible to anyone. Since it's an internal itowns thing, I prefer to keep this private.

File: src/Process/3dTilesProcessing.js (hunk 7 above)
Comment (peppsac) [reply src/Process/3dTilesProcessing.js]: Yeah.. or I could remove camera (since it's part of context as well).

File: src/Core/MainLoop.js (hunk 8 above)
Comment (peppsac) [reply src/Core/MainLoop.js:191]: I don't understand, can you rephrase?

File: src/Process/PointCloudProcessing.js (hunk 9 above)
Comment (peppsac) [reply src/Process/PointCloudProcessing.js:175]: Err.. I didn't get this comment either. Can you reformulate?

File: src/Core/MainLoop.js (hunk 6 above)
Comment (gchoqueux) [reply src/Core/MainLoop.js:189]: it's possible to move it in Camera and it's not accessible

File: src/Core/MainLoop.js (hunk 3 above)
Comment (gchoqueux) [reply src/Core/MainLoop.js]: ok, comment on your approach

File: src/Core/MainLoop.js (hunk 8 above)
Comment (gchoqueux) [reply src/Core/MainLoop.js:191]: `view.camera.camera3D.near ` it 's the distance between the near plane and the camera's position, on camera's  z Axis.

`context.distance.min` is the distance between camera's position and bbox. This distance should be reprojected on camera's  z Axis.

File: src/Process/3dTilesProcessing.js (hunk 7 above)
Comment (gchoqueux) [reply src/Process/3dTilesProcessing.js]: it's preferable to pass only the parameters used in the function

File: src/Process/3dTilesProcessing.js [hunk 10]
```diff
@@ -257,24 +257,36 @@ export function pre3dTilesUpdate(context, layer) {
...
     } else if (node.boundingVolume.box) {
         // boundingVolume.box is affected by matrixWorld
         boundingVolumeBox.copy(node.boundingVolume.box);
         boundingVolumeBox.applyMatrix4(node.matrixWorld);
         node.distance = boundingVolumeBox.distanceToPoint(camera.camera3D.position);
+        // update distance min/max
+        const s = node.boundingVolume.box.getSize();
+        const maxComponent = Math.max(s.x, Math.max(s.y, s.z));
+        context.distance.min = Math.min(context.distance.min, node.distance);
+        context.distance.max = Math.max(context.distance.max, node.distance + maxComponent);
```
Comment (gchoqueux) [inline src/Process/3dTilesProcessing.js]: `maxComponent` is the diagonal

File: src/Process/3dTilesProcessing.js (hunk 4 above)
Comment (gchoqueux) [reply src/Process/3dTilesProcessing.js]: in line 293

File: src/Process/PointCloudProcessing.js (hunk 5 above)
Comment (gchoqueux) [reply src/Process/PointCloudProcessing.js:89]: it should be factorized then or use the [aggregation bboxs](https://github.com/iTowns/itowns/pull/810#pullrequestreview-139481142) to make the generic method

File: examples/3dtiles.html (hunk 1 above)
Comment (peppsac) [reply examples/3dtiles.html]: I'll restore `GlobeControls`.

File: src/Process/PointCloudProcessing.js (hunk 9 above)
Comment (gchoqueux) [reply src/Process/PointCloudProcessing.js:175]: ~this is important for now I will come back to it later~

File: src/Core/MainLoop.js (hunk 3 above)
Comment (peppsac) [reply src/Core/MainLoop.js]: Done

File: src/Process/3dTilesProcessing.js (hunk 4 above)
Comment (peppsac) [reply src/Process/3dTilesProcessing.js]: Nope, because in the third case ther's a bounding sphere, not box.
I've simplified this part anyway.

File: src/Process/3dTilesProcessing.js (hunk 10 above)
Comment (peppsac) [reply src/Process/3dTilesProcessing.js]: Makes sense, I'll change this.

File: src/Process/PointCloudProcessing.js (hunk 5 above)
Comment (peppsac) [reply src/Process/PointCloudProcessing.js:89]: Done.

File: src/Process/PointCloudProcessing.js (hunk 9 above)
Comment (peppsac) [reply src/Process/PointCloudProcessing.js:175]: > this is important for now I will come back to it later

:confused:

**Comment by peppsac** [comment]
@gchoqueux: I pushed an update that should fix all your comments (except [the mysterious one](https://github.com/iTowns/itowns/pull/810#discussion_r204467818))

File: src/Process/PointCloudProcessing.js (hunk 9 above)
Comment (gchoqueux) [reply src/Process/PointCloudProcessing.js:175]: this is not important for now I will come back to it later

File: src/Core/MainLoop.js (hunk 6 above)
Comment (peppsac) [reply src/Core/MainLoop.js:189]: Then give me an example please :)

File: src/Core/MainLoop.js (hunk 8 above)
Comment (peppsac) [reply src/Core/MainLoop.js:191]: `min` is the minimum distance between the camera's center and any (visible) objects in the scene.
So I don't see why / what should be reprojected on camera's z axis.

Can you make a drawing of a case where using the `min` distance as the `near` is incorrect?

**Comment by nosy-b** [comment]
@gchoqueux What do you think now?

**Comment by gchoqueux** [comment]
Bouding volumes should be aggregated to obtain the scene's bbox (only with the displayed objects). And at the end, calculate the distance from the camera to this box
